// cortml/lib/loader/index.js
'use strict';


var loader = require('./loader');
var dumper = require('./dumper');


function deprecated(name) {
  return function () {
    throw new Error('Function ' + name + ' is deprecated and cannot be used.');
  };
}


module.exports.Type                = require('./type');
module.exports.Schema              = require('./schema');
module.exports.FAILSAFE_SCHEMA     = require('./schema/failsafe');
module.exports.JSON_SCHEMA         = require('./schema/json');
module.exports.CORE_SCHEMA         = require('./schema/core');
module.exports.DEFAULT_SAFE_SCHEMA = require('./schema/default_safe');
module.exports.DEFAULT_FULL_SCHEMA = require('./schema/default_full');
module.exports.load                = loader.load;
module.exports.loadAll             = loader.loadAll;
module.exports.safeLoad            = loader.safeLoad;
module.exports.safeLoadAll         = loader.safeLoadAll;
module.exports.dump                = dumper.dump;
module.exports.safeDump            = dumper.safeDump;
module.exports.YAMLException       = require('.//exception');

// Deprecated schema names from JS-YAML 2.0.x
module.exports.MINIMAL_SCHEMA = require('./schema/failsafe');
module.exports.SAFE_SCHEMA    = require('./schema/default_safe');
module.exports.DEFAULT_SCHEMA = require('./schema/default_full');

// Deprecated functions from JS-YAML 1.x.x
module.exports.scan           = deprecated('scan');
module.exports.parse          = deprecated('parse');
module.exports.compose        = deprecated('compose');
module.exports.addConstructor = deprecated('addConstructor');

module.exports.TYPES = [
  // Failsafe
  require('./type/str'),
  require('./type/seq'),
  require('./type/map'),

  // JSON
  require('./type/null'),
  require('./type/bool'),
  require('./type/int'),
  require('./type/float'),

  // Default
  require('./type/timestamp'),
  require('./type/merge'),
  require('./type/binary'),
  require('./type/omap'),
  require('./type/pairs'),
  require('./type/set'),

  // JS
  require('./type/js/undefined'),
  require('./type/js/regexp'),
  require('./type/js/function'),
];
